eeps/eep-0073.md

```diff
@@ -0,0 +1,212 @@
+side of `<-` or `<=`. Generators have the following form: 
+
+    Pattern <- List                                  %% List Generator
+    Pattern <= Bitstring                             %% Bitstring Generator
+    Pattern_1 := Pattern_2 <- Map                    %% Map Generator
+
+All qualifiers and filters can be freely used and mixed in all 3 kinds of
+comprehensions. The following example shows a list comprehension with a
+list generator and a bitstring generator.
+
+    [{X,Y} || X <- [1,2,3], <<Y>> <= <<4,5,6>>].
+
+This EEP proposes the addition of zip generators. A zip generator is two or
+more generators connected by `&&`. Zip generators is constructed to connect
+any number of the 3 kinds of generators above. Zip generators can be used
+in list, binary, or map comprehensions in the same way.
+
+For example, if the two generators in the above example is combined together
+as a zip generator, the comprehension would look like:
+
+    [{X,Y} || X <- [1,2,3] && <<Y>> <= <<4,5,6>>].
+
+For every zip generator of the form
+`G1 && ... && Gn`, it is evaluated to have the same result as `zip/n` where
+
+    zip([H1|T1], ..., [Hn|Tn]) ->
+        [{H1,...,Hn} | zip(T1, ..., Tn)];
+    zip([], ..., []) ->
+        [].
+
+Therefore, the above comprehension evaluates to `[{1,4}, {2,5}, {3,6}]`, which
+is the same as if using `lists:zip/2`.
+
+Zip generator can also be used when a comprehension contains other non-zip
+generators and/or filters. The `&&` symbol has a higher precedence than `,`.
+
+The following example evaluates to `[{b,4}, {c,6}]`. The element `{a,2}` is
+filtered out from the resulting list.
+
+    [{X, Y} || X <- [a, b, c] && <<Y>> <= <<2, 4, 6>>, Y =/= 2].
+
+Comparing to using helper functions, there is one advantage of using a zip
+generator: The Erlang compiler does not generate any tuple when a zip
+generator is translated into core Erlang. The generated code reflects the
+programmer's intent, which is to collect one element from every list at a
+time without creating a list of tuples.
+
+Error Behaviors
+================
+
+One would expect that when errors happen, a zip generator behaves the same
+as `lists:zip/2`, `lists:zip3/3`, and also the `zip/n` function above when
+more than 3 lists are zipped together. The design and implementation of
+zip generators aim to achieve that both for compiled code and for comprehensions
+evaluated in Erlang shell.
+
+Generators of Different Lengths
+--------------
+
+`lists:zip/2` and `lists:zip3/3` will fail if the given lists are not of the
+same length, where `zip/n` will also crash. Therefore, a zip generator raises a
+`bad generators` error when it discovers that the given generators are of
+different lengths. 
+
+When a zip generator crashes because the containing generators are of
+different lengths, the error message is a tuple, where the first element
+is the atom `bad_generators`, and the second element is a tuple that contains
```

Comment:
Nitpick: I believe all of the errors raised by Erlang do not have a `_` after `bad`. You have `badarg`, `badfun`, `badkey`, `badmap`, etc. I would probably keep it consistent.

Reply:
That's my bad. I should write the user-facing error message, not just the internal one. Some errors in `erl_error.erl` use `_` though, like `bad_generator`, but it's internal. Thanks for pointing it out :)

Reply:
Ah, if we already have `bad_generator`, then `bad_generators` make the most sense. Thank you!